crates/pathfinder/src/consensus/inner/fetch_validators.rs

```diff
@@ -0,0 +1,134 @@
+use anyhow::Context;
+use pathfinder_common::{BlockHeader, BlockId, ChainId, ContractAddress};
+use pathfinder_consensus::{PublicKey, SigningKey, Validator, ValidatorSet};
+use pathfinder_storage::Storage;
+use rand::rngs::OsRng;
+
+use crate::config::ConsensusConfig;
+
+#[derive(Clone)]
+pub struct L2ValidatorSetProvider {
+    storage: Storage,
+    chain_id: ChainId,
+    config: ConsensusConfig,
+}
+
+impl L2ValidatorSetProvider {
+    pub fn new(storage: Storage, chain_id: ChainId, config: ConsensusConfig) -> Self {
+        Self {
+            storage,
+            chain_id,
+            config,
+        }
+    }
+
+    fn get_block_header(&self) -> anyhow::Result<BlockHeader> {
+        let mut db_conn = self
+            .storage
+            .connection()
+            .context("Failed to create database connection")?;
+        let db_tx = db_conn
+            .transaction()
+            .context("Failed to create database transaction")?;
+        let block_header = db_tx
+            .block_header(BlockId::Latest)
+            .context("Querying latest block header")?;
+        block_header.ok_or_else(|| anyhow::anyhow!("No block header found"))
+    }
+}
+
+impl pathfinder_consensus::ValidatorSetProvider<ContractAddress> for L2ValidatorSetProvider {
+    fn get_validator_set(
+        &self,
+        height: u64,
+    ) -> Result<ValidatorSet<ContractAddress>, anyhow::Error> {
+        fetch_validators(
+            &self.storage,
+            self.chain_id,
+            self.get_block_header()?,
+            height,
+            &self.config,
+        )
+    }
+}
+
+// TODO:
+//
+// Currently, the validator fetching functionality lives in its own crate
+// (validator-fetcher) because we have a temporary internal RPC method that we
+// use for convenient testing.
+//
+// This separation allows us to easily expose and test the functionality through
+// the RPC while the specification for validator fetching is still being
+// finalized.
+//
+// Once we have a final spec, the functionality from the validator-fetcher crate
+// will be migrated into this file and the temporary crate (along with its RPC
+// method) will be removed.
```

Comment:
1. How would we test this functionality once the RPC method is gone? Using a unit test?
2. Is there any test that we can run rn in the CI/locally? I don't see it in the PR...